### PR TITLE
Optimize TypeConversionExtensions.ConvertTo

### PR DESCRIPTION
**Pull request by eerhardt**
- Change GetTypeConverterTypeName to return the Type directly, so we aren't calling Type.AssemblyQualifiedName, and then Type.GetType.
- Cache TypeConverters from the "toType". This saves allocating toType.CustomAttributes, and allocating a new TypeConverter every time this is called with the same type, e.g. Color.

## Benchmark Results

The easiest public API I could find to exercise this path was Setter.ProvideValue. But this shows up in a startup profile for AppThemeBindingExtension.ProvideValue:

image

### Before

|       Method |     Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|---------:|---------:|-------:|------:|------:|----------:|
| ProvideValue | 10.71 us | 0.189 us | 0.232 us | 0.5035 |     - |     - |   3.11 KB |

### After

|       Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ProvideValue | 309.4 ns | 4.65 ns | 4.13 ns | 0.0353 |     - |     - |     224 B |

### CODE REVIEW

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs [hunk 1]
```diff
@@ -42,24 +43,24 @@ static class TypeConversionExtensions
 			{ typeof(Font), typeof(FontTypeConverter) }
 		};
 
-		private static Dictionary<Type, Func<TypeConverter>> KnownConverterFactories = new()
+		static Dictionary<Type, Func<TypeConverter>> KnownConverterFactories = new()
 		{
 			{ typeof(Font), () => new FontTypeConverter() }
```
Comment (jonathanpeppers) [inline src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:48]: General question for @StephaneDelcroix, should we have a lot more converter types in this dictionary? I'm thinking about types like `Color`, `Thickness`, etc.

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs [hunk 2]
```diff
@@ -42,24 +43,24 @@ static class TypeConversionExtensions
 			{ typeof(Font), typeof(FontTypeConverter) }
 		};
 
-		private static Dictionary<Type, Func<TypeConverter>> KnownConverterFactories = new()
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:45]: this is music to my ears

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:48]: @jonathanpeppers glad you asked

well, at least we shouldn't have multiple caches... TypeConversionExtensions.KnownConverters, TCE.KnownConverterFactories, BindableProperty.KnownTypeConverter; and then we have NodeILExtensions.KnonwnCompiledTypeConverters

Type Conversion happens at multiple places:
- when setting a value from xaml
- when applying a binding (this includes AppThemeBinding, DynamicResource, ...)
- when applying a style

ColorTypeConverter and ThicknessTypeConverter do _not_ need to be cached in a known list, as they are attributed with a TypeConverterAttribute, they need to be in the KnownCOMPILEDTypeConverter as they're not attributed with the ProvideCompiledAttibute.

The only converter that can not be retrieved using reflection is UriTypeConverter as System.Uri is NOT attributed on .NET core 6 (even though System provides a UriTypeConverter that we should use now that we replaced our own base TypeConverter type by System.ComponentModel.TypeConverter. Side note, System.Uri is attributed in the net framework profile, go figure). so why do we whitelist ColorTypeConverter and EasingTypeConverter in BindableProperty.cs is something that need clarification.

I don't recall the details of why FontTypeConverter need to be whitelisted, but I recall that we almost got rid of Font as a property type in most (all?) of our controls (because compound properties are a mess and produce too many propertychange events, and redraw).

We could/should cache the TypeConverters:
- at runtime when we encounter them
- at startup time using reflection (fuck no)
- or at compile time using a sourcegen
Our own TypeConverters are known beforehand, but I'm not keen on maintaining a list of those (I'd prefer a sourcegen), but there's also the TypeConverters defined in 3rd party assemblies. Oh, and caching type converters isn't a solution for avoiding reflection, as TypeConverterAttributes on properties take priority over the ones defined on Type, but we could cache that too (using the PropertyInfo as key e.g.)

To summarise this digression on TypeConverters, here's a list of action points:
- [ ] Investigate the usage of `Font` in Controls. It's probably unused, in that case `FontTypeConverter` and the 2 caches in TypeConversionExtension could go away
- [ ] Investigate `Font` usage in Core, and the reason it was ported there, and not deprecated
- [ ] Investigate why ColorTypeConverter and EasingTypeConverter are whitelisted
- [ ] Investigate caching TypeConverters for properties and types. I'd recommend going for a sourcegen solution and generate a switch instead of a dictionary to avoid allocation bump
- [ ] None of our TypeConverters have any side effect, and instances could be reused (we can speak for 3rd party converters, unfortunately...)

Comment (StephaneDelcroix) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:48]: We need to track the progress of this in a different issue...

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs [hunk 3]
```diff
@@ -71,42 +72,51 @@ static class TypeConversionExtensions
...
-					converterTypeName = memberInfo.CustomAttributes.GetTypeConverterTypeName() ?? converterTypeName;
-				if (converterTypeName == null && KnownConverterFactories.TryGetValue(toType, out var converterFactory))
+				if (KnownConverterFactories.TryGetValue(toType, out var converterFactory))
 					return converterFactory();
-				if (converterTypeName == null)
-					return null;
-				var convertertype = Type.GetType(converterTypeName);
-				return (TypeConverter)Activator.CreateInstance(convertertype);
+
+				if (!s_converterCache.TryGetValue(toType, out TypeConverter converter))
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:86]: The converters we define and provide are reuse-safe and do not have any side effects. We can not guarantee for 3rd party converters.

There are 2 options here:
- only cache our own converters
- require converters to be reusable (I'd go for this)

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs [hunk 4]
```diff
@@ -71,42 +72,51 @@ static class TypeConversionExtensions
 		{
 			Func<TypeConverter> getConverter = () =>
 			{
-				MemberInfo memberInfo;
+				if (minfoRetriever != null && minfoRetriever() is MemberInfo memberInfo)
+				{
+					if (memberInfo.CustomAttributes.GetTypeConverterType() is Type converterType)
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:77]: I'd add a second cache level here with a ConcurrentDictionary<MemberInfo, TypeConverter> with the double benefit of not having to cycle through CustomAttributes and avoid re-creating those converters.

Also, we should cache the fact that no Converter has been found for that MemberInfo to fallback faster, and avoid iterating through the CustomAttributes

Comment (StephaneDelcroix) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:77]: As mentioned in a different comment, these 2 caches should be generated at build time using a source generator, but this isn't a blocker for this PR

**Review by StephaneDelcroix — CHANGES_REQUESTED** [review]

File: src/Controls/src/Core/Xaml/TypeConversionExtensions.cs (hunk 3 above)
Comment (eerhardt) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:86]: My understanding is that TypeConverters are meant to be cached and reused. They don't have state on them.

[WPF caches them](https://github.com/dotnet/wpf/blob/89d172db0b7a192de720c6cfba5e28a1e7d46123/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Markup/XamlTypeMapper.cs#L3478-L3491).

And so does System.ComponentModel:
1. [It stores them on the ReflectedTypeData class](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.ReflectedTypeData.cs#L199-L225)
2. [Which is stored in a ReflectTypeDescriptionProvider](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L909-L914)
3. [Which is stored in a static (weak) hashtable](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L1433-L1434)

Comment (eerhardt) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs:86]: At one point, I was considering using the rest of `System.ComponentModel` to get the converter via `TypeDescriptor.GetConverter`. But I felt this was too heavy-weight to bring in all of the `TypeDescriptor` infrastructure.

Comment (StephaneDelcroix) [reply src/Controls/src/Core/Xaml/TypeConversionExtensions.cs]: if 3rd party converters are stageful, it's garbage in, garbage out. works for me

**Review by StephaneDelcroix — APPROVED** [review]
